### PR TITLE
Dynarray.{equal, compare}

### PR DESCRIPTION
**Pull request by gasche**
This was suggested by @OlivierNicole in https://github.com/ocaml/ocaml/pull/12885#issuecomment-2067761369 .

Having comparison functions for `Dynarray` is useful because they do not have a canonical representation: two dynarrays may have the same elements, but yet have `(=)` return false or `Stdlib.compare` return non-0 due to having a different capacity.

I reused and adapted the docstring of List.equal and Seq.compare.

## Comparing of different-length arrays

For `compare` there is a choice between:
1. what wikipedia calls the lexicographic order: `[10] > [1; 2]` because `10 > 1`
2. what wikipedia calls the "shortlex" order: `[10] < [1; 2]` because the first list is shorter (before comparing the elements)

Using `Stdlib.compare (Dynarray.to_list a) (Dynarray.to_list b)` would result in lexicographic order (1), and `Stdlib.compare (Dynarray.to_array a) (Dynarray.to_array b)` would result in shortlex order (2). (In other words, "doing the same as other data structures" does not force us to choose one or the other.)

I went for the shortlex order which is slightly faster.

### CODE REVIEW

File: stdlib/dynarray.ml [hunk 1]
```diff
@@ -891,6 +891,50 @@ let filter_map f a =
...
+          cmp
+            (unsafe_get arr1 ~dummy:dum1 ~i ~length)
+            (unsafe_get arr2 ~dummy:dum2 ~i ~length)
+        in
+        if c <> 0 then c
+        else loop (i + 1)
+    in
+    let r = loop 0 in
+    check_same_length "equal" a1 ~length;
+    check_same_length "equal" a2 ~length;
```
Comment (yallop) [inline stdlib/dynarray.ml]:
```suggestion
    check_same_length "compare" a1 ~length;
    check_same_length "compare" a2 ~length;
```

Comment (gasche) [reply stdlib/dynarray.ml]: Thanks!

File: stdlib/dynarray.mli [hunk 2]
```diff
@@ -308,6 +308,31 @@ val filter_map : ('a -> 'b option) -> 'a t -> 'b t
...
+    and for each pair of elements [ai], [bi] at the same
+    position we have [eq ai bi].
+
+    @since 5.3
+*)
+
+val compare : ('a -> 'a -> int) -> 'a t -> 'a t -> int
+(** Provided the function [cmp] defines a preorder on elements,
+    [compare cmp a b] compares the lengths of [a] and [b], and
+    then their elements according to the lexicographic preorder.
```
Comment (dbuenzli) [inline stdlib/dynarray.mli]: I don't think the formulation here captures well that lexicographic only happens if the arrays are of equal length.

Perhaps something like `[compare cmp a b] compares first [a] and [b] by their length, and then, if equal, by their elements…`

File: stdlib/dynarray.mli [hunk 3]
```diff
@@ -308,6 +308,31 @@ val filter_map : ('a -> 'b option) -> 'a t -> 'b t
...
+
+    Comparison functions iterate over their two dynarray inputs; it is
+    a programming error to change their length during the iteration,
+    see the {{!section:iteration} Iteration} section above.
+ *)
+
+val equal : ('a -> 'a -> bool) -> 'a t -> 'a t -> bool
+(** [equal eq a b] holds when [a] and [b] have the same length,
+    and for each pair of elements [ai], [bi] at the same
+    position we have [eq ai bi].
```
Comment (dbuenzli) [inline stdlib/dynarray.mli]: Note that position has a specific meaning in OCaml (see the `String` module). I would rather use index (there's a few places in the `.mli` where index should also be substituted in my opinion). Also I would simply use `get` to index: 

```
(** [equal eq a b] holds when [a] and [b] have the same length,
    and for all indices [i] we have [eq (get a i) (get b i)]. *)
```

File: stdlib/dynarray.mli [hunk 4]
```diff
@@ -308,6 +308,31 @@ val filter_map : ('a -> 'b option) -> 'a t -> 'b t
     ignoring strings that cannot be converted to integers.
 *)
 
+(** {1:comparison Comparison functions}
+
+    Comparison functions iterate over their two dynarray inputs; it is
```
Comment (dbuenzli) [inline stdlib/dynarray.mli]: I found over `their two dynarray inputs` a bit hard to parse, the  `two` can be dropped, and perhaps even the `dynarray`, the context seems quite clear. I would go with the first of the following, but the others are fine too:

```
Comparison functions iterate over their arguments
Comparison functions iterate over their dynarray arguments
Comparison functions iterate over their inputs
Comparison functions iterate over their dynarray inputs
```

**Review by dbuenzli — COMMENTED** [review]
I just left bikesheedding comments on the docstrings.

**Comment by gasche** [comment]
Thanks both, I took your comments into account.

**Comment by yallop** [comment]
> Having comparison functions for `Dynarray` is useful because they do not have a canonical representation: two dynarrays may have the same elements, but yet have `(=)` return false or `Stdlib.compare` return non-0 due to having a different capacity.

Extending this reasoning, we should have comparison functions for other "container" types in the standard library, too, because their elements might not have canonical representations.  For example, `Array.equal` is also useful, because `(=)` might return false for an `int Dynarray.t Array.t` comparison where corresponding `Dynarray.t` elements have the same `int` elements but different capacities.

**Comment by gasche** [comment]
I don´t disagree and in fact I was surprised to find that `Array` does not have those functions. (We have `equal` functions for base types thanks to #84.) But I would prefer to restrict the scope of the present PR to Dynarray comparisons only to avoid getting stuck.

File: stdlib/dynarray.ml [hunk 5]
```diff
@@ -891,6 +891,50 @@ let filter_map f a =
...
+    r
+  end
+
+let compare cmp a1 a2 =
+  let Pack {arr = arr1; length = length; dummy = dum1} = a1 in
+  let Pack {arr = arr2; length = len2; dummy = dum2} = a2 in
+  if length <> len2 then length - len2
```
Comment (OlivierNicole) [inline stdlib/dynarray.ml:918]: Smart!

**Review by OlivierNicole — APPROVED** [review]
This looks good to me.

File: stdlib/dynarray.ml (hunk 5 above)
Comment (gasche) [reply stdlib/dynarray.ml:918]: (... and only correct because we know that they are both non-negative, I think.)

Comment (OlivierNicole) [reply stdlib/dynarray.ml:918]: Well, insofar as you consider that a dynarray of length −12 is lesser than a dynarray of length −2, it is correct!

**Comment by shindere** [comment]
Olivier Nicole (2024/05/16 08:45 -0700):
> Well, insofar as you consider that a dynarray of length −12 is lesser than a dynarray of length −2, it is correct!

:-)
